src/zcash/NoteEncryption.cpp

```diff
@@ -0,0 +1,169 @@
+#include "NoteEncryption.hpp"
+#include <stdexcept>
+#include "sodium.h"
+#include <boost/static_assert.hpp>
+#include "prf.h"
+
+#define NOTEENCRYPTION_CIPHER_KEYSIZE 32
+
+void clamp_curve25519(unsigned char key[crypto_scalarmult_SCALARBYTES])
+{
+    key[0] &= 248;
+    key[31] &= 127;
+    key[31] |= 64;
+}
+
+void KDF(unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE],
+    const uint256 &dhsecret,
+    const uint256 &epk,
+    const uint256 &pk_enc,
+    const uint256 &hSig,
+    unsigned char nonce
+   )
+{
+    if (nonce == 0xff) {
+        throw std::logic_error("no additional nonce space for KDF");
+    }
```

Comment:
Nitpick: This should technically allow `0xff` to be used as a nonce (if we're implementing noteencryption as abstract as possible, so that it could be used for >2-airity pours in the future). If we aren't then maybe we should test `nonce >= 2` to make sure we never call `encrypt()` more than twice. I'd rather see this logic in `encrypt()` where the nonce gets incremented, too.